module5/lessons/Week1/ErrorHandling.md

```diff
@@ -0,0 +1,326 @@
+
+If some value could possibly be null (even if you don't think it will be), it's best to add a check and specify what you want to happen in that null case.
+
+For example, we could add a null check to our view like this:
+```c#
+@model Movie
+
+@if (Model == null)
+{
+    <p>No movie details available.</p>
+}
+else
+{
+    <h1>Movie Details</h1>
+
+    <p>Title: @Model.Title</p>
+
+    <p>Genre: @Model.Genre</p>
+}
+```
+
+#### In The Controller
+
+Take a look at this action from the `MoviesController.cs`.
+
+```c#
+ // GET: /Movies/<id>
+[Route("Movies/{id:int}")]
+public IActionResult Show(int id)
+{
+    var movie = _context.Movies.Find(id);
+    return View(movie);
+}
+```
+
+<section class="call-to-action" markdown="1">
+With your partner: Discuss what might be null here? What might happen if those values are null?
+</section>
+
+Let's add some checks to make sure those values are not null.
+
+```c#
+[Route("Movies/{id:int}")]
+public IActionResult Show(int id)
+{
+    
+    if (id == null)
+    {
+        return NotFound();
+    }
+
+    var movie = _context.Movies.Find(id);
+    if (movie == null)
+    {
+        return NotFound();
+    }
+
+    return View(movie);
+}
+```
+
+This is much less likely to break, you will also hear this referred to as more "robust code".
+
+<section class="call-to-action" markdown="1">
+With your partner: Discuss if you think it's better to have null checks in the controller, in the view, or both?
+</section>
+
+<aside class="instructor-notes" markdown="1">
+    <p><strong>Instructor Note</strong><br>
+    My opinion is it's best to have error handling in both, it doesn't hurt to defend against all the weird cases. But it feels more important to have error handling in the controller. Ideally then you wouldn't even send some unexpected data to the view. This goes back to the error handling principle of handling the error as close to the cause of the weirdness as possible.</p>
+</aside>
+
+## Data Validations
+
+What would happen if I put an incredibly long title in for a movie? At some very very very large point it might be to large for our SQL database and throw an error, and even before that a massive title has the potential to mess up the usability of our website.
+
+Or what would happen if someone input a string for a field that can only store an integer? Unless we have some error handling, that would cause an error when we try to write to the database.
+
+The way around these and many other potential problems is to add **Data Validations**. Data validations are checks and rules that are applied to data to ensure its accuracy, integrity, and consistency. The purpose of data validations is to prevent bad data from entering a system, database, or application.
+
+The goal of our data validation is to prevent writing invalid data to the database and instead show users a helpful error message.
+
+![.NET Error](/assets/images/module5/week1/CannotExceed100Chars.png)
+
+
+### Data Validation Step 1: The Model

```

Comment:
This lesson in general feels like it leans more towards sit and get than interactive. I was trying to think about adding more try it out during the lesson, but then I think it would be too long. So I'm thinking we power through these examples and discussions in the morning and then have lots of lab try it out practice time in the afternoon

Reply:
I agree - I think with these topics it would be hard to get students to self-research the key topics we want them to get.  I think as long as the lab is pretty exploratory, we should be good here.  Also students will have just come off of intermission, so that might be the best time for them to be able to power through some sit-n-get.